### PR TITLE
cli: drop support for passing comma-separated template parameters through --param/--value

### PR DESCRIPTION
**Pull request by mmilata**
Related to #10687, #10952.

TODO:
- [x] Extend to `--env` parameters
- [x] Rebase pflag in kubernetes (kubernetes/kubernetes#35700)
- [x] Docs update (openshift/openshift-docs#3107 + openshift/openshift-docs#3123)
- [ ] Public notification after merge - mailing lists, release notes

The bump commit was created manually because I've been unable to make `godep restore` work for me. The bump is needed in order to have https://godoc.org/github.com/spf13/pflag#StringArray.

### CODE REVIEW

**Comment by liggitt** [comment]
can we bump to the same level as 1.4 upstream (1560c1005499d61b80f865c04d39ca7505bf7f0b)? https://github.com/kubernetes/kubernetes/blob/release-1.4/Godeps/Godeps.json#L2050-L2053

**Comment by liggitt** [comment]
or if this is post-1.4, to whatever is currently in master upstream

**Comment by liggitt** [comment]
ah, nevermind... looks like 1.4 didn't have stringarray

**Comment by mmilata** [comment]
@liggitt kubernetes master has c7e63cf4530bcd3ba943729cee0efeff2ebea63f which has StringArray but is missing a bunch of small fixes

**Comment by mmilata** [comment]
@bparees @smarterclayton PTAL

Docs update: openshift/openshift-docs#3107.

Does [test] run the cmd tests?

**Comment by mmilata** [comment]
> Does [test] run the cmd tests?
It does. [test] again

**Comment by liggitt** [comment]
Please open an upstream PR bumping to the same version then

File: pkg/cmd/cli/cmd/newapp.go [hunk 1]
```diff
@@ -166,7 +166,7 @@ func NewCmdNewApplication(name, baseName string, f *clientcmd.Factory, out, erro
 	cmd.Flags().StringSliceVar(&config.Templates, "template", config.Templates, "Name of a stored template to use in the app.")
 	cmd.Flags().StringSliceVarP(&config.TemplateFiles, "file", "f", config.TemplateFiles, "Path to a template file to use for the app.")
 	cmd.MarkFlagFilename("file", "yaml", "yml", "json")
-	cmd.Flags().StringSliceVarP(&config.TemplateParameters, "param", "p", config.TemplateParameters, "Specify a list of key value pairs (e.g., -p FOO=BAR,BAR=FOO) to set/override parameter values in the template.")
+	cmd.Flags().StringArrayVarP(&config.TemplateParameters, "param", "p", config.TemplateParameters, "Specify a key-value pair (e.g., -p FOO=BAR) to set/override parameter value in the template.")
```
Comment (bparees) [inline pkg/cmd/cli/cmd/newapp.go]: "to set/override *a* parameter value"

File: pkg/cmd/cli/cmd/process.go [hunk 2]
```diff
@@ -57,20 +57,20 @@ var (
...
 		Run: func(cmd *cobra.Command, args []string) {
-			err := RunProcess(f, out, cmd, args)
+			err := RunProcess(f, out, errout, cmd, args)
 			kcmdutil.CheckErr(err)
 		},
 	}
 	cmd.Flags().StringP("filename", "f", "", "Filename or URL to file to read a template")
 	cmd.MarkFlagFilename("filename", "yaml", "yml", "json")
-	cmd.Flags().StringSliceP("value", "v", nil, "Specify a list of key-value pairs (eg. -v FOO=BAR,BAR=FOO) to set/override parameter values")
+	cmd.Flags().StringArrayP("value", "v", nil, "Specify a key-value pair (eg. -v FOO=BAR) to set/override parameter value in the template.")
```
Comment (bparees) [inline pkg/cmd/cli/cmd/process.go]: "to set/override *a* parameter value"

File: pkg/cmd/cli/cmd/newapp.go [hunk 3]
```diff
@@ -166,7 +166,7 @@ func NewCmdNewApplication(name, baseName string, f *clientcmd.Factory, out, erro
...
 	cmd.Flags().StringSliceVar(&config.Groups, "group", config.Groups, "Indicate components that should be grouped together as <comp1>+<comp2>.")
 	cmd.Flags().StringSliceVarP(&config.Environment, "env", "e", config.Environment, "Specify key-value pairs of environment variables to set into each container. This doesn't apply to objects created from a template, use parameters instead.")
```
Comment (bparees) [inline pkg/cmd/cli/cmd/newapp.go]: we should do the same thing for env (drop comma support, add a warning)

**Comment by mmilata** [comment]
@bparees updated to handle `--env` arguments as well. Multiline values shouldn't be a problem, `StringArray` should handle them fine, `StringSlice` does not. There's a [test] case for this in `test/cmd/process.sh`.

File: pkg/bootstrap/docker/up.go [hunk 4]
```diff
@@ -139,7 +139,7 @@ func NewCmdUp(name, fullName string, f *osclientcmd.Factory, out io.Writer) *cob
 	cmd.Flags().StringVar(&config.HostDataDir, "host-data-dir", "", "Directory on Docker host for OpenShift data. If not specified, etcd data will not be persisted on the host.")
 	cmd.Flags().BoolVar(&config.PortForwarding, "forward-ports", config.PortForwarding, "Use Docker port-forwarding to communicate with origin container. Requires 'socat' locally.")
 	cmd.Flags().IntVar(&config.ServerLogLevel, "server-loglevel", 0, "Log level for OpenShift server")
-	cmd.Flags().StringSliceVarP(&config.Environment, "env", "e", config.Environment, "Specify key value pairs of environment variables to set on OpenShift container")
+	cmd.Flags().StringArrayVarP(&config.Environment, "env", "e", config.Environment, "Specify a key value pair of environment variable to set on OpenShift container")
```
Comment (bparees) [inline pkg/bootstrap/docker/up.go]: "Specify a key value pair of environment variable"->"Specify a key value pair for an environment variable"

File: pkg/cmd/cli/cmd/newapp.go [hunk 5]
```diff
@@ -166,9 +166,9 @@ func NewCmdNewApplication(name, baseName string, f *clientcmd.Factory, out, erro
...
 	cmd.MarkFlagFilename("file", "yaml", "yml", "json")
-	cmd.Flags().StringSliceVarP(&config.TemplateParameters, "param", "p", config.TemplateParameters, "Specify a list of key value pairs (e.g., -p FOO=BAR,BAR=FOO) to set/override parameter values in the template.")
+	cmd.Flags().StringArrayVarP(&config.TemplateParameters, "param", "p", config.TemplateParameters, "Specify a key-value pair (e.g., -p FOO=BAR) to set/override a parameter value in the template.")
 	cmd.Flags().StringSliceVar(&config.Groups, "group", config.Groups, "Indicate components that should be grouped together as <comp1>+<comp2>.")
-	cmd.Flags().StringSliceVarP(&config.Environment, "env", "e", config.Environment, "Specify key-value pairs of environment variables to set into each container. This doesn't apply to objects created from a template, use parameters instead.")
+	cmd.Flags().StringArrayVarP(&config.Environment, "env", "e", config.Environment, "Specify a key-value pair of environment variable to set into each container. This doesn't apply to objects created from a template, use parameters instead.")
```
Comment (bparees) [inline pkg/cmd/cli/cmd/newapp.go]: Specify a key value pair of environment variable -> Specify a key value pair for an environment variable

File: pkg/cmd/cli/cmd/newbuild.go [hunk 6]
```diff
@@ -128,7 +128,7 @@ func NewCmdNewBuild(name, baseName string, f *clientcmd.Factory, in io.Reader, o
 	cmd.Flags().StringVar(&config.Name, "name", "", "Set name to use for generated build artifacts.")
 	cmd.Flags().StringVar(&config.To, "to", "", "Push built images to this image stream tag (or Docker image repository if --to-docker is set).")
 	cmd.Flags().BoolVar(&config.OutputDocker, "to-docker", false, "Have the build output push to a Docker repository.")
-	cmd.Flags().StringSliceVarP(&config.Environment, "env", "e", config.Environment, "Specify key value pairs of environment variables to set into resulting image.")
+	cmd.Flags().StringArrayVarP(&config.Environment, "env", "e", config.Environment, "Specify a key value pair of environment variable to set into resulting image.")
```
Comment (bparees) [inline pkg/cmd/cli/cmd/newbuild.go]: Specify a key value pair of environment variable -> Specify a key value pair for an environment variable

File: pkg/cmd/cli/cmd/set/env.go [hunk 7]
```diff
@@ -93,7 +93,7 @@ func NewCmdEnv(fullName string, f *clientcmd.Factory, in io.Reader, out, errout
 	cmd.Flags().StringP("containers", "c", "*", "The names of containers in the selected pod templates to change - may use wildcards")
 	cmd.Flags().StringP("from", "", "", "The name of a resource from which to inject enviroment variables")
 	cmd.Flags().StringP("prefix", "", "", "Prefix to append to variable names")
-	cmd.Flags().StringSliceVarP(&env, "env", "e", env, "Specify key value pairs of environment variables to set into each container.")
+	cmd.Flags().StringArrayVarP(&env, "env", "e", env, "Specify a key value pair of environment variable to set into each container.")
```
Comment (bparees) [inline pkg/cmd/cli/cmd/set/env.go]: Specify a key value pair of environment variable -> Specify a key value pair for an environment variable

File: pkg/cmd/cli/cmd/startbuild.go [hunk 8]
```diff
@@ -90,7 +90,7 @@ func NewCmdStartBuild(fullName string, f *clientcmd.Factory, in io.Reader, out,
 		},
 	}
 	cmd.Flags().StringVar(&o.LogLevel, "build-loglevel", o.LogLevel, "Specify the log level for the build log output")
-	cmd.Flags().StringSliceVarP(&o.Env, "env", "e", o.Env, "Specify key value pairs of environment variables to set for the build container.")
+	cmd.Flags().StringArrayVarP(&o.Env, "env", "e", o.Env, "Specify a key value pair of environment variable to set for the build container.")
```
Comment (bparees) [inline pkg/cmd/cli/cmd/startbuild.go]: Specify a key value pair of environment variable -> Specify a key value pair for an environment variable

File: pkg/cmd/util/cmd.go [hunk 9]
```diff
@@ -163,3 +163,11 @@ func VersionedPrintObject(fn func(*cobra.Command, meta.RESTMapper, runtime.Objec
 		return fn(c, mapper, obj, out)
 	}
 }
+
+func WarnAboutCommaSeparation(errout io.Writer, values []string, flag string) {
+	for _, value := range values {
+		if strings.ContainsRune(value, ',') {
+			fmt.Fprintf(errout, "warning: %s no longer accepts comma-separated list of values. %q will be treated as a single key-value pair.\n", flag, value)
```
Comment (bparees) [inline pkg/cmd/util/cmd.go]: s/list/lists/

Comment (bparees) [inline pkg/cmd/util/cmd.go]: we should only warn if the value looks like "key1=value1,key2=value2".  In other words, we have to see both a command and an equal after key1= for it to be a problem.  There's nothing wrong with "key1=a,b,c"

**Comment by bparees** [comment]
@mfojtik you will need to review this also since it's changing oc set for deploymentconfigs.

**Comment by mmilata** [comment]
@bparees fixed the descriptions, changed the warning functions to match against `.*=.*,.*=.*`.

File: test/cmd/newapp.sh [hunk 10]
```diff
@@ -63,6 +63,31 @@ os::cmd::expect_success_and_text 'oc new-app ruby-helloworld-sample -o yaml' 'AD
...
+# check that warning is printed when --env PARAM1=VAL1,PARAM2=VAL2 is used
+os::cmd::expect_success_and_text 'oc new-app php --env PASS=one,two=three -o yaml' 'no longer accepts comma-separated list'
+# check that warning is not printed when --param/env doesn't contain two k-v pairs
+os::cmd::expect_success_and_not_text 'oc new-app php --env DEBUG=disabled -o yaml' 'no longer accepts comma-separated list'
+os::cmd::expect_success_and_not_text 'oc new-app php --env LEVELS=INFO,WARNING -o yaml' 'no longer accepts comma-separated list'
+os::cmd::expect_success_and_not_text 'oc new-app ruby-helloworld-sample --param MYSQL_USER=mysql -o yaml' 'no longer accepts comma-separated list'
+os::cmd::expect_success_and_not_text 'oc new-app ruby-helloworld-sample --param MYSQL_PASSWORD=com,ma -o yaml' 'no longer accepts comma-separated list'
+# check that warning is not printed when parameters/env are passed positionally
+os::cmd::expect_success_and_not_text 'oc new-app php PASS=one,two=three -o yaml' 'no longer accepts comma-separated list'
+os::cmd::expect_success_and_not_text 'oc new-app ruby-helloworld-sample MYSQL_PASSWORD=one,two=three -o yaml' 'no longer accepts comma-separated list'
```
Comment (bparees) [inline test/cmd/newapp.sh]: these two tests should also confirm that the parameter/env is not split.

File: test/cmd/newapp.sh [hunk 11]
```diff
@@ -63,6 +63,31 @@ os::cmd::expect_success_and_text 'oc new-app ruby-helloworld-sample -o yaml' 'AD
...
+
+# new-build
+# check that env vars are not split on commas and warning is printed where they previously have
+os::cmd::expect_success_and_text 'oc new-build --binary php --env X=Y,Z=W -o yaml' 'value: Y,Z=W'
+os::cmd::expect_success_and_text 'oc new-build --binary php --env X=Y,Z=W -o yaml' 'no longer accepts comma-separated list'
+os::cmd::expect_success_and_not_text 'oc new-build --binary php --env X=Y,Z,W -o yaml' 'no longer accepts comma-separated list'
+os::cmd::expect_success_and_not_text 'oc new-build --binary php --env X=Y -o yaml' 'no longer accepts comma-separated list'
+os::cmd::expect_success_and_not_text 'oc new-build --binary php X=Y,Z=W -o yaml' 'no longer accepts comma-separated list'
```
Comment (bparees) [inline test/cmd/newapp.sh]: confirm this test does not split the env into two.

File: test/cmd/process.sh [hunk 12]
```diff
@@ -55,5 +55,18 @@ os::cmd::expect_failure_and_text "oc process -f '${required_params}' --value=req
...
 os::cmd::expect_failure_and_text "oc process -f '${required_params}' --value=required_param=someval --labels======" 'error parsing labels'
 
+# values are not split on commas, required parameter is not recognized
+os::cmd::expect_failure_and_text "oc process -f '${required_params}' --value=optional_param=a,required_param=b" 'parameter required_param is required and must be specified'
+# warning is printed iff --value looks like two k-v pairs separated by comma
+os::cmd::expect_success_and_text "oc process -f '${required_params}' --value=required_param=a,b=c,d" 'no longer accepts comma-separated list'
+os::cmd::expect_success_and_not_text "oc process -f '${required_params}' --value=required_param=a_b_c_d" 'no longer accepts comma-separated list'
+os::cmd::expect_success_and_not_text "oc process -f '${required_params}' --value=required_param=a,b,c,d" 'no longer accepts comma-separated list'
+# warning is not printed for template values passed as positional arguments
+os::cmd::expect_success_and_not_text "oc process -f '${required_params}' required_param=a,b=c,d" 'no longer accepts comma-separated list'
```
Comment (bparees) [inline test/cmd/process.sh]: confirm argument is not split.

File: test/cmd/templates.sh [hunk 13]
```diff
@@ -48,8 +45,10 @@ os::cmd::expect_success_and_text 'oc process ADMIN_USERNAME=myuser ADMIN_PASSWOR
 os::cmd::expect_success_and_text 'oc process -f test/templates/testdata/guestbook.json ADMIN_USERNAME=myuser ADMIN_PASSWORD=mypassword'       '"mypassword"'
 # Argument values with commas are honored
 os::cmd::expect_success 'oc create -f examples/sample-app/application-template-stibuild.json'
-os::cmd::expect_success_and_text 'oc process ruby-helloworld-sample MYSQL_USER=myself MYSQL_PASSWORD=my,1%pass'  '"myself"'
-os::cmd::expect_success_and_text 'oc process MYSQL_USER=myself MYSQL_PASSWORD=my,1%pass ruby-helloworld-sample'  '"my,1%pass"'
+os::cmd::expect_success_and_text 'oc process ruby-helloworld-sample MYSQL_USER=myself MYSQL_PASSWORD=my,1%pa=s'        '"myself"'
+os::cmd::expect_success_and_text 'oc process MYSQL_USER=myself MYSQL_PASSWORD=my,1%pa=s ruby-helloworld-sample'        '"my,1%pa=s"'
```
Comment (bparees) [inline test/cmd/templates.sh]: looks like this covers at least one of the test suggestions i made above.

I wonder if cmd/templates.sh and cmd/process.sh should be combined into a single test script so we can do some de-duping.

Not for this PR though.

**Comment by mmilata** [comment]
Following assertion failed because `oc new-app` never supported template parameters as positional arguments (unlike `oc process`):
```
os::cmd::expect_success_and_text 'oc new-app ruby-helloworld-sample MYSQL_PASSWORD=one,two=three -o yaml' 'value: one,two=three'
```
Pushed new version with the test removed.

**Comment by danmcp** [comment]
Flaked on https://github.com/openshift/origin/issues/10988

**Comment by mmilata** [comment]
[test]

**Comment by mmilata** [comment]
The test got stuck on
```
[k8s.io] EmptyDir volumes
/data/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go:793
  should support (root,0644,default) [Conformance]
  /data/src/github.com/openshift/origin/vendor/k8s.io/kubernetes/test/e2e/common/empty_dir.go:97
```
Should be reported as a flake?

[test] again?

**Comment by smarterclayton** [comment]
report the test run as a flake in an issue (might be a new docker failure
mode)

[test]

**Comment by mmilata** [comment]
Filed #11671

**Comment by mmilata** [comment]
Flake #11406

**Comment by mmilata** [comment]
[test]

**Comment by mmilata** [comment]
Flake #10988 or #10773 or both. [test]

**Comment by mmilata** [comment]
@bparees, modified the tests as you suggested, PTAL

**Comment by bparees** [comment]
lgtm [merge]

**Comment by mmilata** [comment]
@smarterclayton you requested a public notification for this change in https://github.com/openshift/origin/pull/10952#issuecomment-253033055

Where do I add a release note? I'm planning to send an email to users@ list as well.

**Comment by smarterclayton** [comment]
In openshift-docs (there should be a 3.4 release note list)

**Pull request merged**